<!-- Case 1 -->
<launch>

	<!-- genco相关参数 -->
	<arg name="collision_therahold" default="40.0"/><!-- 碰撞判定距离 -->
	<arg name="safe_therahold" default="1.5"/><!-- 飞机安全距离 -->
	<arg name="total_time" default="10.0"/><!-- 飞行时间 -->
	<arg name="ros_t" default="0.2"/><!-- 间隔时间 -->


	<!-- 无人机集群数量 -->
	<arg name="swarm_num_uav" default="4"/>
	<arg name="uav1_id" default="1"/>
	<arg name="uav2_id" default="2"/>
	<arg name="uav3_id" default="3"/>
	<arg name="uav4_id" default="4"/>
	<!-- 无人机初始位置和终点位置 -->
	<arg name="uav1_init_x" default="5.0"/>
    <arg name="uav1_init_y" default="5.0"/>
    <arg name="uav1_init_z" default="2.9"/>
    <!-- <arg name="uav1_init_z" default="20.0"/> -->

	<arg name="uav1_end_x" default="-20.0"/>
    <arg name="uav1_end_y" default="-19.0"/>
    <arg name="uav1_end_z" default="39.0"/>

    <arg name="uav2_init_x" default="5.0"/>
    <arg name="uav2_init_y" default="-5.0"/>
    <arg name="uav2_init_z" default="3.2"/>
    <!-- <arg name="uav2_init_z" default="20.0"/> -->

	<arg name="uav2_end_x" default="-19.0"/>
    <arg name="uav2_end_y" default="21.0"/>
    <arg name="uav2_end_z" default="38.5"/>

    <arg name="uav3_init_x" default="-5.0"/>
    <arg name="uav3_init_y" default="-5.0"/>
    <arg name="uav3_init_z" default="2.8"/>
    <!-- <arg name="uav3_init_z" default="20.0"/> -->

	<arg name="uav3_end_x" default="19.5"/>
    <arg name="uav3_end_y" default="20.5"/>
    <arg name="uav3_end_z" default="10.5"/>

    <arg name="uav4_init_x" default="-5.0"/>
    <arg name="uav4_init_y" default="5.0"/>
    <arg name="uav4_init_z" default="3.1"/>
    <!-- <arg name="uav4_init_z" default="20.0"/> -->

	<arg name="uav4_end_x" default="21.0"/>
    <arg name="uav4_end_y" default="-19.5"/>
    <arg name="uav4_end_z" default="10.5"/>

	<!-- swarm_estimator 参数 -->
	<arg name="input_source" default="2"/>
	<!-- swarm_controller 参数 -->
	<arg name="collision_flag" default="0"/>
	<arg name="Takeoff_height" default="1.0"/>
	<arg name="Disarm_height" default="0.2"/>
	<arg name="Land_speed" default="0.2"/>
	<arg name="Land_mode" default="1"/>
	<!-- 地理围栏 -->
	<arg name="x_min" default="-100.0"/>
	<arg name="x_max" default="100.0"/>
	<arg name="y_min" default="-100.0"/>
	<arg name="y_max" default="100.0"/>
	<arg name="z_min" default="-1.0"/>
	<arg name="z_max" default="50.0"/>

	<!-- 启动顺序 -->
	<!-- 启动SITL,并加载模型 -->
	<!-- 启动 swarm_estimator -->
	<!-- 启动 swarm_controller -->
	<!-- 启动 地面站 -->
	<!-- 启动 阵型变化程序 -->
	<!-- 启动 Gazebo -->

    <!-- 1号飞机 -->
	<include file="$(find genco)/launch/sitl_single_uav.launch">
		<arg name="swarm_num_uav" value="$(arg swarm_num_uav)"/>
		<arg name="uav_id" value="$(arg uav1_id)"/>
		<arg name="sdf" value="$(find prometheus_gazebo)/amov_models/Multisolo/uav1/uav1.sdf"/>
		<arg name="model" value="uav1"/>
		<arg name="uav_init_x" value="$(arg uav1_init_x)"/>
		<arg name="uav_init_y" value="$(arg uav1_init_y)"/>
		<arg name="uav_init_z" value="0.15"/>
		<arg name="uav_init_yaw" value="0.0"/>
	</include>

    <!-- 2号飞机 -->
	<include file="$(find genco)/launch/sitl_single_uav.launch">
		<arg name="swarm_num_uav" value="$(arg swarm_num_uav)"/>
		<arg name="uav_id" value="$(arg uav2_id)"/>
		<arg name="sdf" value="$(find prometheus_gazebo)/amov_models/Multisolo/uav2/uav2.sdf"/>
		<arg name="model" value="uav2"/>
		<arg name="uav_init_x" value="$(arg uav2_init_x)"/>
		<arg name="uav_init_y" value="$(arg uav2_init_y)"/>
		<arg name="uav_init_z" value="0.15"/>
		<arg name="uav_init_yaw" value="0.0"/>
	</include>

    <!-- 3号飞机 -->
	<include file="$(find genco)/launch/sitl_single_uav.launch">
		<arg name="swarm_num_uav" value="$(arg swarm_num_uav)"/>
		<arg name="uav_id" value="$(arg uav3_id)"/>
		<arg name="sdf" value="$(find prometheus_gazebo)/amov_models/Multisolo/uav3/uav3.sdf"/>
		<arg name="model" value="uav3"/>
		<arg name="uav_init_x" value="$(arg uav3_init_x)"/>
		<arg name="uav_init_y" value="$(arg uav3_init_y)"/>
		<arg name="uav_init_z" value="0.15"/>
		<arg name="uav_init_yaw" value="0.0"/>
	</include>

    <!-- 4号飞机 -->
	<include file="$(find genco)/launch/sitl_single_uav.launch">
		<arg name="swarm_num_uav" value="$(arg swarm_num_uav)"/>
		<arg name="uav_id" value="$(arg uav4_id)"/>
		<arg name="sdf" value="$(find prometheus_gazebo)/amov_models/Multisolo/uav4/uav4.sdf"/>
		<arg name="model" value="uav4"/>
		<arg name="uav_init_x" value="$(arg uav4_init_x)"/>
		<arg name="uav_init_y" value="$(arg uav4_init_y)"/>
		<arg name="uav_init_z" value="0.15"/>
		<arg name="uav_init_yaw" value="0.0"/>
	</include>

	<!-- run the swarm_estimator.cpp -->
	<node pkg="prometheus_swarm_control" type="swarm_estimator" name="swarm_estimator_uav1" output="screen">
		<param name="uav_id" value="$(arg uav1_id)"/>
		<param name="input_source" value="$(arg input_source)" />
	</node>
	<node pkg="prometheus_swarm_control" type="swarm_estimator" name="swarm_estimator_uav2" output="screen">
		<param name="uav_id" value="$(arg uav2_id)"/>
		<param name="input_source" value="$(arg input_source)" />
	</node>
	<node pkg="prometheus_swarm_control" type="swarm_estimator" name="swarm_estimator_uav3" output="screen">
		<param name="uav_id" value="$(arg uav3_id)"/>
		<param name="input_source" value="$(arg input_source)" />
	</node>
	<node pkg="prometheus_swarm_control" type="swarm_estimator" name="swarm_estimator_uav4" output="screen">
		<param name="uav_id" value="$(arg uav4_id)"/>
		<param name="input_source" value="$(arg input_source)" />
	</node>

	
	<!-- 启动swarm_controller -->
	<node pkg="prometheus_swarm_control" type="swarm_controller" name="swarm_controller_uav1" output="screen">
		<param name="swarm_num_uav" value="$(arg swarm_num_uav)"/>
		<param name="uav_id" value="$(arg uav1_id)"/>
		<param name="collision_flag" value="$(arg collision_flag)"/>
		<param name="Takeoff_height" value="$(arg Takeoff_height)"/>
		<param name="Disarm_height" value="$(arg Disarm_height)"/>
		<param name="Land_speed" value="$(arg Land_speed)"/>
		<param name="Land_mode" value="$(arg Land_mode)"/>
		<param name="geo_fence/x_min" value="$(arg x_min)"/>
		<param name="geo_fence/x_max" value="$(arg x_max)"/>
		<param name="geo_fence/y_min" value="$(arg y_min)"/>
		<param name="geo_fence/y_max" value="$(arg y_max)"/>
		<param name="geo_fence/z_min" value="$(arg z_min)"/>
		<param name="geo_fence/z_max" value="$(arg z_max)"/>
	</node>

	<node pkg="prometheus_swarm_control" type="swarm_controller" name="swarm_controller_uav2" output="screen">
		<param name="swarm_num_uav" value="$(arg swarm_num_uav)"/>
		<param name="uav_id" value="$(arg uav2_id)"/>
		<param name="collision_flag" value="$(arg collision_flag)"/>
		<param name="Takeoff_height" value="$(arg Takeoff_height)"/>
		<param name="Disarm_height" value="$(arg Disarm_height)"/>
		<param name="Land_speed" value="$(arg Land_speed)"/>
		<param name="Land_mode" value="$(arg Land_mode)"/>
		<param name="geo_fence/x_min" value="$(arg x_min)"/>
		<param name="geo_fence/x_max" value="$(arg x_max)"/>
		<param name="geo_fence/y_min" value="$(arg y_min)"/>
		<param name="geo_fence/y_max" value="$(arg y_max)"/>
		<param name="geo_fence/z_min" value="$(arg z_min)"/>
		<param name="geo_fence/z_max" value="$(arg z_max)"/>
	</node>

	<node pkg="prometheus_swarm_control" type="swarm_controller" name="swarm_controller_uav3" output="screen">
		<param name="swarm_num_uav" value="$(arg swarm_num_uav)"/>
		<param name="uav_id" value="$(arg uav3_id)"/>
		<param name="collision_flag" value="$(arg collision_flag)"/>
		<param name="Takeoff_height" value="$(arg Takeoff_height)"/>
		<param name="Disarm_height" value="$(arg Disarm_height)"/>
		<param name="Land_speed" value="$(arg Land_speed)"/>
		<param name="Land_mode" value="$(arg Land_mode)"/>
		<param name="geo_fence/x_min" value="$(arg x_min)"/>
		<param name="geo_fence/x_max" value="$(arg x_max)"/>
		<param name="geo_fence/y_min" value="$(arg y_min)"/>
		<param name="geo_fence/y_max" value="$(arg y_max)"/>
		<param name="geo_fence/z_min" value="$(arg z_min)"/>
		<param name="geo_fence/z_max" value="$(arg z_max)"/>
	</node>

	<node pkg="prometheus_swarm_control" type="swarm_controller" name="swarm_controller_uav4" output="screen">
		<param name="swarm_num_uav" value="$(arg swarm_num_uav)"/>
		<param name="uav_id" value="$(arg uav4_id)"/>
		<param name="collision_flag" value="$(arg collision_flag)"/>
		<param name="Takeoff_height" value="$(arg Takeoff_height)"/>
		<param name="Disarm_height" value="$(arg Disarm_height)"/>
		<param name="Land_speed" value="$(arg Land_speed)"/>
		<param name="Land_mode" value="$(arg Land_mode)"/>
		<param name="geo_fence/x_min" value="$(arg x_min)"/>
		<param name="geo_fence/x_max" value="$(arg x_max)"/>
		<param name="geo_fence/y_min" value="$(arg y_min)"/>
		<param name="geo_fence/y_max" value="$(arg y_max)"/>
		<param name="geo_fence/z_min" value="$(arg z_min)"/>
		<param name="geo_fence/z_max" value="$(arg z_max)"/>
	</node>

	
	<!-- 阵型控制程序 -->
	<node pkg="genco" type="start_formation" name="start_formation" output="screen" launch-prefix="gnome-terminal --tab --">
		<param name="swarm_num_uav" value="$(arg swarm_num_uav)"/>
		<param name="sim_mode" value="true"/>

		<param name="init_pos_1_x" value="$(arg uav1_init_x)"/>
		<param name="init_pos_1_y" value="$(arg uav1_init_y)"/>
		<param name="init_pos_1_z" value="$(arg uav1_init_z)"/>

		<param name="init_pos_2_x" value="$(arg uav2_init_x)"/>
		<param name="init_pos_2_y" value="$(arg uav2_init_y)"/>
		<param name="init_pos_2_z" value="$(arg uav2_init_z)"/>

		<param name="init_pos_3_x" value="$(arg uav3_init_x)"/>
		<param name="init_pos_3_y" value="$(arg uav3_init_y)"/>
		<param name="init_pos_3_z" value="$(arg uav3_init_z)"/>

		<param name="init_pos_4_x" value="$(arg uav4_init_x)"/>
		<param name="init_pos_4_y" value="$(arg uav4_init_y)"/>
		<param name="init_pos_4_z" value="$(arg uav4_init_z)"/>
	</node>

	<!-- 启动地面站 -->
	<node pkg="prometheus_swarm_control" type="swarm_ground_station" name="swarm_ground_station" output="screen" launch-prefix="gnome-terminal --tab --">	
		<param name="swarm_num_uav" value="$(arg swarm_num_uav)"/>
	</node>


	<!-- 启动genco节点 -->
	<node pkg="genco" type="genco_node" name="genco_node_1" output="screen" launch-prefix="gnome-terminal --tab --">
		<param name="swarm_num_uav" value="$(arg swarm_num_uav)"/>
		<param name="self_id" value="$(arg uav1_id)"/>

		<param name="init_pos_x" value="$(arg uav1_init_x)"/>
		<param name="init_pos_y" value="$(arg uav1_init_y)"/>
		<param name="init_pos_z" value="$(arg uav1_init_z)"/>

		<!-- <param name="end_pos_x" value="$(arg uav1_end_x)"/>
		<param name="end_pos_y" value="$(arg uav1_end_y)"/>
		<param name="end_pos_z" value="$(arg uav1_end_z)"/> -->
		<param name="end_pos_x" value="$(arg uav3_init_x)"/>
		<param name="end_pos_y" value="$(arg uav3_init_y)"/>
		<param name="end_pos_z" value="$(arg uav3_init_z)"/>

		<param name="collision_therahold" value="$(arg collision_therahold)"/>
		<param name="safe_therahold" value="$(arg safe_therahold)"/>

		<param name="ros_t" value="$(arg ros_t)"/>
		<param name="total_time" value="$(arg total_time)"/>
	</node>

	<node pkg="genco" type="genco_node" name="genco_node_2" output="screen" launch-prefix="gnome-terminal --tab --">
		<param name="swarm_num_uav" value="$(arg swarm_num_uav)"/>
		<param name="self_id" value="$(arg uav2_id)"/>

		<param name="init_pos_x" value="$(arg uav2_init_x)"/>
		<param name="init_pos_y" value="$(arg uav2_init_y)"/>
		<param name="init_pos_z" value="$(arg uav2_init_z)"/>

		<!-- <param name="end_pos_x" value="$(arg uav2_end_x)"/>
		<param name="end_pos_y" value="$(arg uav2_end_y)"/>
		<param name="end_pos_z" value="$(arg uav2_end_z)"/> -->
		<param name="end_pos_x" value="$(arg uav4_init_x)"/>
		<param name="end_pos_y" value="$(arg uav4_init_y)"/>
		<param name="end_pos_z" value="$(arg uav4_init_z)"/>

		<param name="collision_therahold" value="$(arg collision_therahold)"/>
		<param name="safe_therahold" value="$(arg safe_therahold)"/>

		<param name="ros_t" value="$(arg ros_t)"/>
		<param name="total_time" value="$(arg total_time)"/>

	</node>
		<param name="end_pos_x" value="$(arg uav4_init_x)"/>

	<node pkg="genco" type="genco_node" name="genco_node_3" output="screen" launch-prefix="gnome-terminal --tab --">
		<param name="swarm_num_uav" value="$(arg swarm_num_uav)"/>
		<param name="self_id" value="$(arg uav3_id)"/>

		<param name="init_pos_x" value="$(arg uav3_init_x)"/>
		<param name="init_pos_y" value="$(arg uav3_init_y)"/>
		<param name="init_pos_z" value="$(arg uav3_init_z)"/>

		<!-- <param name="end_pos_x" value="$(arg uav3_end_x)"/>
		<param name="end_pos_y" value="$(arg uav3_end_y)"/>
		<param name="end_pos_z" value="$(arg uav3_end_z)"/> -->
		<param name="end_pos_x" value="$(arg uav1_init_x)"/>
		<param name="end_pos_y" value="$(arg uav1_init_y)"/>
		<param name="end_pos_z" value="$(arg uav1_init_z)"/>

		<param name="collision_therahold" value="$(arg collision_therahold)"/>
		<param name="safe_therahold" value="$(arg safe_therahold)"/>

		<param name="ros_t" value="$(arg ros_t)"/>
		<param name="total_time" value="$(arg total_time)"/>

	</node>

	<node pkg="genco" type="genco_node" name="genco_node_4" output="screen" launch-prefix="gnome-terminal --tab --">
		<param name="swarm_num_uav" value="$(arg swarm_num_uav)"/>
		<param name="self_id" value="$(arg uav4_id)"/>

		<param name="init_pos_x" value="$(arg uav4_init_x)"/>
		<param name="init_pos_y" value="$(arg uav4_init_y)"/>
		<param name="init_pos_z" value="$(arg uav4_init_z)"/>

		<!-- <param name="end_pos_x" value="$(arg uav4_end_x)"/>
		<param name="end_pos_y" value="$(arg uav4_end_y)"/>
		<param name="end_pos_z" value="$(arg uav4_end_z)"/> -->
		<param name="end_pos_x" value="$(arg uav2_init_x)"/>
		<param name="end_pos_y" value="$(arg uav2_init_y)"/>
		<param name="end_pos_z" value="$(arg uav2_init_z)"/>

		<param name="collision_therahold" value="$(arg collision_therahold)"/>
		<param name="safe_therahold" value="$(arg safe_therahold)"/>

		<param name="ros_t" value="$(arg ros_t)"/>
		<param name="total_time" value="$(arg total_time)"/>

	</node>

	<!-- 启动rviz,设为false可关闭 -->
	<arg name="visualization" default="true"/>
	<group if="$(arg visualization)">
        <node type="rviz" name="rviz" pkg="rviz" args="-d $(find genco)/view.rviz" />
    </group>

	<!-- 启动Gazebo -->
	<!-- Gazebo configs -->
    <arg name="gui" default="true"/>
    <arg name="debug" default="false"/>
    <arg name="verbose" default="false"/>
    <arg name="paused" default="false"/>
    <arg name="respawn_gazebo" default="false"/>
	<arg name="world" default="$(find prometheus_gazebo)/worlds/prometheus_logo.world"/>
    <!-- Gazebo sim -->
    <include file="$(find gazebo_ros)/launch/empty_world.launch">
        <arg name="gui" value="$(arg gui)"/>
        <arg name="debug" value="$(arg debug)"/>
        <arg name="verbose" value="$(arg verbose)"/>
        <arg name="paused" value="$(arg paused)"/>
        <arg name="respawn_gazebo" value="$(arg respawn_gazebo)"/>
        <arg name="world_name" value="$(arg world)"/>
    </include>
</launch>
